master/internal/experiment.go

```diff
@@ -411,6 +411,39 @@ func (e *internalExperiment) PatchTrialState(msg experiment.PatchTrialState) err
 func (e *internalExperiment) SetGroupMaxSlots(msg sproto.SetGroupMaxSlots) {
 	e.mu.Lock()
 	defer e.mu.Unlock()
+	// Only allow max slots changes if it is not specified as an invariant config or enforced as a
+	// constraint.
+
+	w, err := getWorkspaceByConfig(e.activeConfig)
+	if err != nil {
+		log.Warnf("unable to set max slots")
+		return
+	}
+	enforcedMaxSlots, err := configpolicy.GetEnforcedConfig[int](context.TODO(), &w.ID,
+		"invariant_config",
+		"'resources' -> 'max_slots'", model.ExperimentType)
+	if err != nil {
+		log.Warnf("unable to set max slots")
+		return
+	}
+
+	if enforcedMaxSlots != nil {
+		msg.MaxSlots = enforcedMaxSlots
+	}
+
+	maxSlotsLimit, err := configpolicy.GetEnforcedConfig[int](context.TODO(), &w.ID,
+		"constraints",
+		"'resources' -> 'max_slots'", model.ExperimentType)
+	if err != nil {
+		log.Warnf("unable to set max slots")
+		return
+	}
+
+	if enforcedMaxSlots == nil && maxSlotsLimit != nil && msg.MaxSlots != nil &&
+		*msg.MaxSlots > *maxSlotsLimit {
+		log.Warnf("unable to set max slots")
+		return
+	}
```

Comment:
I think this code should be in a separate function, preferably in `configpolicies` package. The main reason is that I prefer to keep "config policy logic" in `configpolicies`. In other words, an experiment API has no need to know how config policies work, that there's invariant_configs and constraints, precedence, etc. Ideally we would be able to change how config policies work by only changing code in `configpolicies` files. 

The other reason is that the logic can be simplified if it's in its own function. The function could return at line 431 and then have no need to check `enforcedMaxSlots == nil` on line 442.

Reply:
Love this idea, and makes sense! moved this work into its own func, great idea!